### ISSUE TITLE
Multiple Instance Learning

### ISSUE THREAD

**Issue by DepthFirstSearch** [fchollet/keras#3415]
I want to create a ConvNet to do Multiple Instance Learning (MIL).

[To summarize MIL](https://en.wikipedia.org/wiki/Multiple_instance_learning), "the training set consists of labeled “bags”, each of which is a collection of unlabeled instances. A bag is positively labeled if at least one instance in it is positive, and is negatively labeled if all instances in it are negative. The goal of the MIL is to predict the labels of new, unseen bags."

To keep things simple, one batch of training data corresponds to one bag. This implies that with a batch size of _M_, one particular bag contains _M_ instances.

Now, the only thing you have to do is to use a suitable loss function. [Zhou et al.](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.87.1254&rep=rep1&type=pdf#page=6&zoom=auto,-109,414) suggests to use function (1) to compute the error of a bag _B_ (and hence a batch):
![cost](https://cloud.githubusercontent.com/assets/20369248/17462763/b4e37a32-5cb7-11e6-9d5b-624e0cdd876e.png)
where _o_j_ is the actual output of instance _j_ and _d_ the desired output of bag _B_.

Quoting from [Zhou et al.](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.87.1254&rep=rep1&type=pdf#page=6&zoom=auto,-109,414), (1) "means that if at least one instance of a positive training bag is perfectly predicted at positive, or all the instances of a negative bag are predicted as negative, then the error on the concerned bag is zero and the weights of the network will not be updated. Otherwise the weights will be updated according to the error on the instance whose corresponding actual output is maximal among all the instances in the bag."

Now it comes to Keras. I'm actually not sure if my implementation of (1) is correct. Also, I'm not sure how Keras deals with the maximum function in backpropagation. As far as I know, the loss function has to be differentiable in order to compute the gradients. But the maximum function is not differentiable. Here is my implementation:

``` python
def mil_squared_error(y_true, y_pred):
    return K.tile(K.square(K.max(y_pred) - K.max(y_true)), 16)
```

where I have a batch size of 16 and my `class_mode` is `'binary'`. There is probably a lot to improve, e.g. I didn't know how to get the length of `y_true` (but should be equal to 16).

My question is: can I work with such a loss function and is there anything I have to consider? How will backpropagation work this function or (1) in general?

**Comment by hellowangqian** [comment]
Hi mate, I'm having a similar problem now, just wondering if you got any solutions?

**Comment by xypan1232** [comment]
Hi, does it work fro MIL using your solution?

**Comment by asalekin** [comment]
Hi, Did you solve the issue of MIL? If possible can you let us know how?

**Comment by DepthFirstSearch** [comment]
Hi, 
no, I didn't spend a lot of time on solving this issue.
Now looking back at the above loss function, I think it's okay if the function returns just a scalar, so:
```python
def mil_squared_error(y_true, y_pred):
    return K.square(K.max(y_pred) - K.max(y_true))
```

**Comment by bw4sz** [comment]
@mhubrich do you know of a code sample for CNN MIL?
